transport: add timeout for writing GOAWAY on http2Client.Close()

aranjans opened the pull request:
fixes [#7314](https://github.com/grpc/grpc-go/issues/7314)
### Problem
In gRPC 1.64, we released a [feature](https://github.com/grpc/grpc-go/pull/7015) "Send GOAWAY to server on Client Transport Shutdown", post which in case of network hang when we are not able to write GOAWAY frame from client, closing connections will take a very long time.

### What does this PR fix
This PR fixes this issue by having a timeout for loopyWriter to write GOAWAY so that http2Client.Close() doesn't wait for too long.

RELEASE NOTES: 

- transport/http2_client: fix waiting on http2Client.Close() in case of network hang for writing GOAWAY.

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,144 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+
+// TestClientCloseTimeoutOnHang verifies that in the event of a graceful
+// client transport shutdown, i.e., clientTransport.Close(), if the conn hung
+// forever, client should still be close itself and do not wait for long.
```
nit: instead of forever, we should explicitly specify the timer deadline

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,144 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server.
+	go func() {
```
looks like this launch server go routine is same as previous goaway test. May be we can move to a common function and call that instead?

purnesh42H commented on internal/transport/http2_client.go:
```diff
@@ -1009,7 +1009,12 @@ func (t *http2Client) Close(err error) {
 	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
 	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
-	<-t.writerDone
+	timer := time.NewTimer(5 * time.Second)
```
we should make the timer deadline configurable or atleast a const for now for better readability

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,144 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	n, err = hc.Conn.Write(b)
+	if n == goAwayFrameSize { // GOAWAY frame
+		writeHangSignal = make(chan struct{})
+		time.Sleep(15 * time.Second)
```
we can probably just wait on the channel instead of sleeping?

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,144 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Wait until server receives the headers and settings frame as part of greet.
+	<-greetDone
+	ct.Close(errors.New("manually closed by client"))
+	defer close(writeHangSignal)
```
how long does this test run? We probably don't want the test to run for long. Instead, it might be good to repro the hang using a channel and override the timer logic in client from test to time out.

purnesh42H replied:
> how long does this test run? We probably don't want the test to run for long. Instead, it might be good to repro the network hang using a channel and override the timer logic in client from test to time out. This way we will have more control over test duration

aranjans replied:
Test running time is equal to the deadline for loopyWriter to exit(which is for now equivalent to 5 seconds).

purnesh42H replied:
yeah we don't want to run for 5 seconds. I suggested an alternate approach. Let's do something similar where we fake the timeout without actually waiting

purnesh42H commented on internal/transport/http2_client.go:
```diff
@@ -59,6 +59,8 @@ import (
 // atomically.
 var clientConnectionCounter uint64
 
+const LoopyWriterTimeout = 5 * time.Second
```
suggestion: GoAwayLoopyWriterTimeout?

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2670,7 +2745,7 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 	// errorCh verifies that desired GOAWAY not received by server
 	errorCh := make(chan error)
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), 2*defaultTestTimeout)
```
let's avoid increasing timeouts and try to keep the default timeout. One quick approach we can do here is overriding the timeout interval to 0 for this test and hang the net conn write forever. Then we will always timeout on write because of timer and test will finish quickly closing the client without sending goaway frame

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -90,6 +90,7 @@ const (
 	invalidHeaderField
 	delayRead
 	pingpong
+	goAwayFrameSize = 42
```
suggestion: these const values seems to be related for an enum. May be better to have separate const for goAwayFrameSize

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2656,10 +2657,84 @@ func TestConnectionError_Unwrap(t *testing.T) {
+	createServer(t, ConnectOptions{})
+}
+
+var writeHangSignal chan struct{}
```
might be good to give some documentation about hangingConn and why is it needed

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2725,7 +2800,7 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 	}()
 
-	ct, err := NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{}, func(GoAwayReason) {})
+	ct, err := NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, connectOptions, func(GoAwayReason) {})
```
why are client things there in createServer? We should probably move only the goroutine part to a new func and call that in the 2 go away tests. Rest everything can remain in the main test

aranjans replied:
Probably better idea to have all the client and server creation in a single place as both of them are common to `Test/ClientSendsAGoAway` and `Test/ClientCloseTimeoutOnHang`. Refactored the func name to `createClientServerConn`.

purnesh42H replied:
i think only server part is common. For client though, one is using hanging conn and another one is using a regular conn. I would just refactor the server part

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2656,10 +2658,87 @@ func TestConnectionError_Unwrap(t *testing.T) {
+	createClientServerConn(t, ConnectOptions{})
+}
+
+var writeHangSignal chan struct{}
```
documentation about `writeHangSignal`

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2656,10 +2658,87 @@ func TestConnectionError_Unwrap(t *testing.T) {
+func (hc *hangingConn) Write(b []byte) (n int, err error) {
+	n, err = hc.Conn.Write(b)
+	if n == goAwayFrameSize { // GOAWAY frame
+		timer := time.NewTimer(GoAwayLoopyWriterTimeout)
```
this timeout should be more than loopy writer timer. In fact, we don't need timer here so that write always hang so that test becomes more predictable

purnesh42H commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,28 @@ func (t *http2Client) Close(err error) {
 	}
 	t.mu.Unlock()
+	var st *status.Status
 	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
```
let's modify the documentation to include the timer part

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2656,10 +2658,89 @@ func TestConnectionError_Unwrap(t *testing.T) {
+// not wait for long.
+func (s) TestClientCloseTimeoutOnHang(t *testing.T) {
+	origGoAwayLoopyTimeout := GoAwayLoopyWriterTimeout
+	GoAwayLoopyWriterTimeout = 0
```
If we are setting time out to 0, we probably don't need hanging conn as it will always time out? Can you try without hanging conn and see if time out still happens? If yes, test will become much more simpler

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2656,10 +2656,33 @@ func TestConnectionError_Unwrap(t *testing.T) {
+	errorCh := createClientServerLoggingGoAway(t)
+	err := <-errorCh
+	if err != nil {
+		t.Errorf("Error receiving the GOAWAY frame: %v", err)
```
if error happens here its not clear whether it happened at client or server. It would be best if you keep the client and server startup code separate or just have the same code duplicated in the test

easwars reviewed (commented):
Please fix the PR title. 

The PR title  should be of the form: `<package>: <what this pr is about>`. `fix` is not a valid package or component within our codebase.

easwars reviewed (commented):
Also, please add a meaningful PR description. It should not be the responsibility of the person landing on this PR to read through the linked issue to get a sense of what is being done in this PR.

See here for some guidance: https://engdoc.corp.google.com/eng/doc/devguide/coding/review/developer/cl-descriptions.md?cl=head

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+// client transport shutdown, i.e., clientTransport.Close(), if the conn hung
+// for LoopyWriterTimeout, client should still be close itself and should
+// not wait for long.
+func (s) TestClientCloseTimeoutOnHang(t *testing.T) {
```
suggestion: `TestClientSendsAGoAwayFrameWriteTimeout`

easwars replied:
How about `TestClientCloseReturnsEarly` or `TestClientCloseReturnsEarlyWhenGoAwayWriteHangs`?

easwars commented on internal/transport/http2_client.go:
```diff
@@ -59,6 +59,8 @@ import (
 // atomically.
 var clientConnectionCounter uint64
 
+var GoAwayLoopyWriterTimeout = time.Second
```
Why is this exported?

easwars replied:
Did some discussion happen on the value for this timeout? `1s` seems quite short to me. If a discussion happened, it would be worth adding a comment here to say why `1s` is considered a reasonable value for this timeout?

aranjans replied:
@easwars Updated this time out to 5s as per discussion with Doug.

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
+	var st *status.Status
+	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the connection.
+	// See https://httpwg.org/specs/rfc7540.html#GOAWAY. It also waits for loopyWriter to
+	// be closed with a timer to avoid the indefinite blocking.
```
This is not *indefinite* in the true sense, right?

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
+	// be closed with a timer to avoid the indefinite blocking.
 	t.controlBuf.put(&goAway{code: http2.ErrCodeNo, debugData: []byte("client transport shutdown"), closeConn: err})
-	<-t.writerDone
+	timer := time.NewTimer(GoAwayLoopyWriterTimeout)
```
I would use `time.After` for simplicity purposes. See: https://pkg.go.dev/time#After.

If you feel that efficiency (in terms of the timer object being garbage collected) is a concern, then you should arrange for the timer to be `Close`ed.

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
-	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the
-	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	var st *status.Status
+	// Per HTTP/2 spec, a GOAWAY frame must be sent before closing the connection.
```
Please ensure that new comment lines are wrapped at 80-cols. See: https://google.github.io/styleguide/go/decisions#comment-line-length

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
+	timer := time.NewTimer(GoAwayLoopyWriterTimeout)
+	select {
+	case <-t.writerDone:
+		// Append info about previous goaway's if there were any, since this may be important
+		// for understanding the root cause for this connection to be closed.
```
Here as well, wrap to 80-cols.

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,29 @@ func (t *http2Client) Close(err error) {
+			st = status.New(codes.Unavailable, err.Error())
+		}
+	case <-timer.C:
+		t.logger.Warningf("timeout waiting for the loopy writer to be closed.")
```
Something more meaningful to the user would be helpful. Maybe
```
t.logger.Warningf("Failed to write a GOAWAY frame as part of connection close after %s. Giving up and closing the transport.", GoAwayLoopyWriterTimeout)
```

easwars commented on internal/transport/http2_client.go:
```diff
@@ -1035,6 +1042,7 @@ func (t *http2Client) Close(err error) {
 		}
 		sh.HandleConn(t.ctx, connEnd)
 	}
+	t.logger.Infof("Closed the client connection")
```
I don't think this log message is very informative.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2656,7 +2656,7 @@ func TestConnectionError_Unwrap(t *testing.T) {
 }
 
-// Test that in the event of a graceful client transport shutdown, i.e.,
+// TestClientSendsAGoAwayFrame verifies that in the event of a graceful client transport shutdown, i.e.,
```
The earlier comment was good enough. If at all you want to change it, s/Test/Tests/ would be more appropriate.

We have gone through this many times. Test comments can simply start with `Tests ...` instead of `TestFooBarAndBang ...` because the former conveys all the information that the latter does as well.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		}
+		defer sconn.Close()
+		if _, err := io.ReadFull(sconn, make([]byte, len(clientPreface))); err != nil {
+			t.Errorf("Error while writing settings ack: %v", err)
```
This is reading the client preface and not writing a settings ack.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		}
+		fr, _ := sfr.ReadFrame()
+		if _, ok := fr.(*http2.SettingsFrame); !ok {
+			t.Errorf("Expected settings frame, got %v", fr)
```
Use `%T` instead of `%v` to print the `type` of the frame received? Here and below.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+			if goAwayFrame.ErrCode == http2.ErrCodeNo {
+				t.Logf("Received goAway frame from client")
+			} else {
+				errorCh <- fmt.Errorf("received unexpected goAway frame: %v", err)
```
What is the expected GoAway frame supposed to contain? Does it make sense to mention that in the error string?

easwars replied:
Also, this could be written as:

```golang
case *http2.GoAwayFrame:
	if fr.ErrCode == http2.ErrCodeNo {
		t.Logf("Received goAway frame from client")
		return
	}
	errorCh <- fmt.Errorf("...")
```

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		}
+	}()
+
+	ct, err := NewClientTransport(ctx, context.Background(), resolver.Address{Addr: lis.Addr().String()}, ConnectOptions{}, func(GoAwayReason) {})
```
Please don't use a background context. Simply pass `ctx` as the second argument, or make a new context with a finite test timeout and pass that. I see that you are following existing tests (which are currently passing a background context), but we want to avoid this pattern going forward.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	}
+	// Wait until server receives the headers and settings frame as part of greet.
+	<-greetDone
+	ct.Close(errors.New("manually closed by client"))
```
It is not clear how the server hang is reproduced here. Could you please clarify with comments.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// greetDone is used to notify when server is done greeting the client.
+	greetDone := make(chan struct{})
+	// errorCh verifies that desired GOAWAY not received by server
+	errorCh := make(chan error)
```
Consider making this channel with a buffer size of `1`.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	}
+	defer lis.Close()
+	// greetDone is used to notify when server is done greeting the client.
+	greetDone := make(chan struct{})
```
What is this greet talking about? Can we have a better name for this?

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Launch the server.
+	go func() {
+		defer close(errorCh)
+		sconn, err := lis.Accept()
```
s/sconn/conn

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,95 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+			t.Errorf("Error while writing settings ack: %v", err)
+			return
+		}
+		sfr := http2.NewFramer(sconn, sconn)
```
s/sfr/framer/

aranjans commented:
@easwars Thanks for the detailed review, I see doing force push doesn't show the changes made post review. I have updated the PR with code changes in reference to all of your suggestions which include naming convention, adding meaningful comments, wrapping the comments to 80 cols, and also replaced the timer.C with time.After.

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+// not wait for long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override goAwayLoopyWriterTimeout to 0 so that we always
+	// time out while writing GOAWAY on client.Close(). This is
```
suggestion: override timer for writing GOAWAY to 0 so that the conn write always times out. It is equivalent of real network hang when conn write for goaway doesn't finish in specified deadline

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	<-serverGreetingDone
+	// ct.Close will try to send the GOAWAY to server and will fail writing
+	// GOAWAY and will eventually close the loopyWriter as
+	// goAwayLoopyWriterTimeout (time out for writing GOAWAY) is set to zero, which is
```
nit: set to 0

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// ct.Close will try to send the GOAWAY to server and will fail writing
+	// GOAWAY and will eventually close the loopyWriter as
+	// goAwayLoopyWriterTimeout (time out for writing GOAWAY) is set to zero, which is
+	// equivalent to network hang scenario.
```
nit: equivalent to network hang scenario when conn write doesn't finish within specified deadline

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	// Launch the server.
+	go func() {
```
Why do we need this custom server implementation?

What is it important for this server to notify when it is done reading the client preface?

Could you please add comments to describe why you are doing whatever you are doing here?

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	}
+	// Wait until server receives the headers and settings frame as part of greet.
+	<-serverGreetingDone
+	// ct.Close will try to send the GOAWAY to server and will fail writing
+	// GOAWAY and will eventually close the loopyWriter as
```
Why will writing the GOAWAY fail?

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,100 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	select {
+	case <-errorCh:
+	case <-ctx.Done():
+		t.Errorf("Context timed out")
```
Test error strings should give enough useful information for someone running the test and looking at the output. `Context timed out` does not tell me anything useful.

aranjans commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,30 @@ func (t *http2Client) Close(err error) {
+			st = status.New(codes.Unavailable, err.Error())
+		}
+	case <-time.After(goAwayLoopyWriterTimeout):
+		t.logger.Warningf("Failed to write a GOAWAY frame as part of connection close after %v. Giving up and closing the transport.", goAwayLoopyWriterTimeout)
```
I believe no as in else case, we wouldn't had sent the GOAWAY as we failed to do so. WDYT?

purnesh42H replied:
I think you still have to set the status code because you are closing the transport. Probably similar to the else part of writer done i.e `code.Unavailable`?

aranjans commented:
@easwars and @purnesh42H thAnks for the review, I have addressed all your comments and updated the PR.

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,81 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	if err != nil {
+		t.Fatalf("failed to open stream: %v", err)
+	}
+	// ct.Close will try to send the GOAWAY to server and will fail writing
```
suggestion: ct.Close will try to send the GOAWAY to server but `conn.Write` will timeout due to goAwayLoopyWriterTimeout being 0. It is equivalent of a network hang when GOAWAY doesn't finish within specified deadline

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,81 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 }
+
+// Test that in the event of a graceful client transport shutdown
+// , i.e., clientTransport.Close(), if the conn hung for
```
suggestion: , i.e., clientTransport.Close(), if the GOAWAY write is not finished within specified time due to network hang, client should still close without waiting for too long

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		defer close(errorCh)
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting: %v", err)
```
This error should be written to the `errorCh`, and we should return early from here in that case.

aranjans replied:
this comment won't be vaild anymore as we updated the test method.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		defer conn.Close()
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(); err != nil {
+			t.Errorf("Error while writing settings %v", err)
```
Same here. Write the error to the error channel.

aranjans replied:
this comment won't be vaild anymore as we updated the test method.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+
+		frame, err := framer.ReadFrame()
+		if err != nil {
+			return
```
And here too.

aranjans replied:
this comment won't be vaild anymore as we updated the test method.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+				t.Logf("Received goAway frame from client")
+				return
+			}
+			errorCh <- fmt.Errorf("received unexpected goAway frame: %v", err)
```
Add a return statement after this too?

aranjans replied:
this comment won't be vaild anymore as we updated the test method.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// finish within specified deadline.
+	ct.Close(errors.New("manually closed by client"))
+	select {
+	case <-errorCh:
```
Check for the value returned on the channel, and if non-nil, fail the test.

aranjans replied:
@easwars I think we shouldn't check for errorCh as server won't be receiving the GOAWAY at all. And we just want to check that client conn close is not waiting on loopy to close for long and doesn't get context timed out.

easwars replied:
If the server is not expected to receive a GOAWAY at all, the server implementation above should send a `nil` value on the error channel when that expectation is met. 

The usual approach in a test would be as follows:
- main test goroutine spawns a goroutine which does some work, and sends its result on a channel
- main test goroutine reads from the channel, and if error value read from the channel is non-nil, it will fail

aranjans replied:
Updated [here](https://github.com/grpc/grpc-go/pull/7371#issuecomment-2232984289)

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+			return
+		}
+		switch fr := frame.(type) {
+		case *http2.GoAwayFrame:
```
Are we guaranteed to get this GOAWAY frame?

In `http2Client.Close`, since the timeout value is set to `0` in this test, the `case <-time.After(goAwayLoopyWriterTimeout)` will fire before `<-t.writerDone`. And in this case, we set `st` and we go ahead and close the underlying `conn`, by calling `t.conn.Close()` on line 1034, and if this happens before the queued GOAWAY frame was written, then the server would never receive the goaway, right? Or am I missing something?

aranjans replied:
@easwars yeah that's correct and here in this test, server will not receive the GOAWAY as we will get the timeout for  loopy. And we just want to test the timeout here, that is loopy writer doesn't wait for long even if there is network hang.

easwars replied:
I don't understand. You have code in here which returns without an error when you receive a GOAWAY with code `ErrCodeNo`, but you error out when receive a GOAWAY with any other code.

aranjans replied:
Updated [here](https://github.com/grpc/grpc-go/pull/7371#issuecomment-2232984289)

easwars reviewed (commented):
Please reply to comments with something meaningful to the reviewers. Simply adding a thumbs-up emoji does not convey useful information to the reviewer.

aranjans commented:
@easwars After having thought for some time, i realized we don't even need to care about the GOAWAY frame, we just need to make sure in case of network hang, client doesn't wait for long and gets timed out. I have made code changes, and updated the PR for the same.

easwars reviewed (commented):
Force pushes are making this review intractable.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,80 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Write will time out due to goAwayLoopyWriterTimeout being
+	// 0. It is equivalent of a network hang when GOAWAY doesn't
+	// finish within specified deadline.
+	ct.Close(errors.New("manually closed by client"))
```
If the call to `ct.Close` hangs (i.e the bug that we are trying to fix manifests itself), then we will get get to the `select` below.

aranjans commented:
@easwars thanks for your review, i have updated the PR and made all the suggestions you gave.

easwars commented:
> @easwars thanks for your review, i have updated the PR and made all the suggestions you gave.

There are a bunch of comments still unaddressed, mostly around the need for a custom server goroutine in this test, and about writing errors from that goroutine onto an error channel instead of calling `t.Errorf`.

easwars commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,71 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 	}
 }
+
+// Test that in the event of a graceful client transport shutdown
```
I think this test can be rewritten as follows:

```golang
// Tests the scenario where a client transport is closed and writing of the
// GOAWAY frame as part of the close does not complete because of a network
// hang. The test verifies that the client transport is closed without waiting
// for too long.
func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
	// Override timer for writing GOAWAY to 0 so that the connection write
	// always times out. It is equivalent of real network hang when conn
	// write for goaway doesn't finish in specified deadline
	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
	goAwayLoopyWriterTimeout = 0
	defer func() {
		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
	}()

	server, ct, cancel := setUp(t, 0, normal)
	defer cancel()
	defer server.stop()

	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
	if _, err := ct.NewStream(ctx, &CallHdr{}); err != nil {
		t.Fatalf("Failed to open stream: %v", err)
	}

	ctClosed := make(chan struct{})
	go func() {
		// ct.Close will try to send a GOAWAY frame to the server but conn.Write
		// will time out due to goAwayLoopyWriterTimeout being 0. It is
		// equivalent of a network hang when GOAWAY doesn't finish within
		// specified deadline.
		ct.Close(errors.New("manually closed by client"))
		close(ctClosed)
	}()

	select {
	case <-ctClosed:
	case <-ctx.Done():
		t.Errorf("Timeout waiting for Close() to complete")
	}
}
```

aranjans replied:
Ok, thAnks. I have made the change and updated the PR.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,34 @@ func (t *http2Client) Close(err error) {
-	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY.
+	// connection. See https://httpwg.org/specs/rfc7540.html#GOAWAY. It
+	// also waits for loopyWriter to be closed with a timer to avoid the
+	// long blocking in case the connection is half closed.
```
"in case the connection is half closed" is incorrect, right?  Isn't it needed in case the connection is blackholed, i.e. TCP is just stuck?

aranjans replied:
Updated

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,34 @@ func (t *http2Client) Close(err error) {
+		}
+	case <-timer.C:
+		st = status.New(codes.Unavailable, err.Error())
+		t.logger.Warningf("Failed to write a GOAWAY frame as part of connection close after %s. Giving up and closing the transport.", goAwayLoopyWriterTimeout)
```
This doesn't seem like a warning to me.  Let's go with info.

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,44 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+// hang. The test verifies that the client transport is closed without waiting
+// for too long.
+func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+	// Override timer for writing GOAWAY to 0 so that the connection write
+	// always times out. It is equivalent of real network hang when conn
+	// write for goaway doesn't finish in specified deadline
```
This feels like a race?  How can we be sure that the loopyWriter doesn't process the item to exit before the timer is even created?  We need to make the Conn hang on demand, not just override the timeout to zero.  I think we can do this pretty easily by wrapping the conn by injecting a dialer, and set a flag for when we want `Write` to just not do anything.

aranjans replied:
Umm.... I initially had the custom dialer with a hang [here](https://github.com/grpc/grpc-go/blob/59604784b6641bf7df8c3f270f8574915bae54ed/internal/transport/transport_test.go#L2734) but that will require the test to run for few seconds which may cause flakiness. Hence, I decided to override the timeout to 0 so that write is always expected to be time out. WDYT?

dfawley replied:
I'm fine with an instant (or short?) timeout, but this doesn't actually make the connection hang, so it doesn't actually test anything, does it?  I.e. this test will pass even before your changes, right?  (Obviously with the exception that it sets a global variable that doesn't exist @ HEAD.)

aranjans replied:
Ah makes sense.  I have updated the logic of custom dialer now, which hangs the connection.

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2746,44 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	}
+
+	ctClosed := make(chan struct{})
+	go func() {
+		// ct.Close will try to send a GOAWAY frame to the server but conn.Write
+		// will time out due to goAwayLoopyWriterTimeout being 0. It is
+		// equivalent of a network hang when GOAWAY doesn't finish within
+		// specified deadline.
+		ct.Close(errors.New("manually closed by client"))
+		close(ctClosed)
+	}()
+
+	select {
+	case <-ctClosed:
+	case <-ctx.Done():
+		t.Errorf("Timeout waiting for Close() to complete")
+	}
```
This feels excessively paranoid.  We should just call `Close` inline and expect that it will return.  Check the elapsed time if you're paranoid that it waited too long.  Otherwise, if you extend this to its logical conclusion, we need to put _every_ function we call inside a goroutine with a select after it. :)

aranjans replied:
Updated

purnesh42H commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2756,85 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	return hc.Conn.Close()
+}
+
+func (hc *hangingConn) LocalAddr() net.Addr {
```
you don't need to override all methods. I think only write and close are probably enough for the test case

dfawley replied:
Yes please delete all these methods that just call the underlying `net.Conn` directly.

dfawley commented on internal/transport/http2_client.go:
```diff
@@ -1006,24 +1008,35 @@ func (t *http2Client) Close(err error) {
+			st = status.New(codes.Unavailable, err.Error())
+		}
+	case <-timer.C:
+		st = status.New(codes.Unavailable, err.Error())
```
Wait.. so if the timer fires we're losing the goaway debugging info.  That's not ideal.

Why did you couple this error logic with the timer/writerDone stuff?  I don't think that's needed.  Can you leave the error handling stuff alone and just replace `<-t.writerDone` with:
```go
select {
	case <-t.writerDone: // success
	case <-timer.C:
		t.logger.Info(...)
}
```

dfawley replied:
This looks much better now, thanks..

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2756,85 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		goAwayLoopyWriterTimeout = origGoAwayLoopyTimeout
+	}()
+
+	isGreetingsDone = atomic.Bool{}
```
Set to `false` instead please.

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2756,85 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	n, err = hc.Conn.Write(b)
+	if isGreetingsDone.Load() == true {
+		// Add a delay which is more than goAwayLoopyWriterTimeout
+		time.Sleep(2 * time.Second)
```
Please don't use a sleep here.  Instead use a channel or `grpcsync.Event` that comes from the dialer.  We can trigger it after the test case passes.

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2756,85 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// always times out. It is equivalent of real network hang when conn
+	// write for goaway doesn't finish in specified deadline
+	origGoAwayLoopyTimeout := goAwayLoopyWriterTimeout
+	goAwayLoopyWriterTimeout = time.Second
```
Make this much shorter please.  1ms?

aranjans replied:
Updated

aranjans commented on internal/transport/transport_test.go:
```diff
@@ -427,8 +428,16 @@ func setUpServerOnly(t *testing.T, port int, sc *ServerConfig, ht hType) *server
-	return setUpWithOptions(t, port, &ServerConfig{}, ht, ConnectOptions{})
+// isGreetingDone verifies that client-server setup is complete
+// for the test.
+var isGreetingsDone = atomic.Bool{}
```
Sure, updated the `s/Greetings/Greeting/`. 
On the later part, `isGreetingDone` is actually part of setUp method which is being used in most test cases and `isGreetingDone` could be useful in future tests for verifying whether the client server handshakes are complete and conn is ready to use. Would it be worth keeping it as it is, WDYT?

dfawley replied:
Globals are bad.  We should be avoiding them unless it's not possible to.  Let's have this test do its own setup or something instead.

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -427,8 +428,16 @@ func setUpServerOnly(t *testing.T, port int, sc *ServerConfig, ht hType) *server
+// for the test.
+var isGreetingDone = atomic.Bool{}
+
+func setUp(t *testing.T, port int, ht hType, options ...ConnectOptions) (*server, *http2Client, func()) {
+	var copts = ConnectOptions{}
+	if len(options) > 0 {
+		copts = options[0]
+	}
+	return setUpWithOptions(t, port, &ServerConfig{}, ht, copts)
 }
```
What is this all about?  You added an options parameter to `setUp`, but we already have `setUpWithOptions`.  I don't get it.

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2756,67 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	hangConn chan struct{}
+}
+
+func (hc *hangingConn) Read(b []byte) (n int, err error) {
+	n, err = hc.Conn.Read(b)
+	return n, err
+}
```
Delete?

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2758,21 +2749,17 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 	n, err = hc.Conn.Write(b)
-	if isGreetingDone.Load() == true {
+	if hc.isGreetingDone.Load() == true {
 		// Hang the Write for more than goAwayLoopyWriterTimeout
```
Remove this timer?  Otherwise it will release on its own and pass on master@HEAD, won't it?

dfawley replied:
Ping, this doesn't seem done.

aranjans replied:
@dfawley After having more thought on this, I think we would need to have timeout for `hc.hangConn`, otherwise [this](https://github.com/grpc/grpc-go/blob/master/internal/transport/http2_client.go#L467) go-routine will be leaked via [this](https://github.com/grpc/grpc-go/blob/master/internal/transport/controlbuf.go#L644) call. So `hc.Conn.Write` will be hang after we made [this](https://github.com/grpc/grpc-go/pull/7371/files#diff-c90a44d2d36c8cc9c90b97b290cce3bdf6a5086bdac1e51e42d0e7b8f415b8b8R2761) write to the conn, which will eventually make the above go-routine to run forever and end up in leaked go-routine.(btw we already have deadline for net.Conn.Write defined [here](https://github.com/grpc/grpc-go/pull/7371/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R988) at the top of ct.Close, but then the current test will test only for [this](https://github.com/grpc/grpc-go/pull/7371/files#diff-5334018444bcb34d15bcb3c8a3050cdf9c8a40fe41305c3a602ea4bfeb0cd178R1022) loopyWriter to timeout).
WDYT?

purnesh42H replied:
Not entirely familiar with which goroutine you are referring to but one suggestion is that you probably don't have to do the actual write `n, err = hc.Conn.Write(b)`? You can just compare the passed bytes `b` if its goaway (like using `string(b)`) and then just hang? That way you don't have any risk of spawning any other go routine (if any).

aranjans replied:
Actually we don't even need to compare it, we can just make sure hangingConn hangs from the time it receives the goAway request. 

@dfawley i have removed the timer logic, and updated the PR.

purnesh42H replied:
probably good to release the `<-hc.hangConn` at the end of test?

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2806,9 +2785,30 @@ func (s) TestClientCloseReturnsEarlyWhenGoAwayWriteHangs(t *testing.T) {
+		if err != nil {
+			return nil, err
+		}
+		isGreetingDone.Store(false)
```
This is unnecessary IIUC, since the dialer is only used once.

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2758,21 +2749,17 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
-	n, err = hc.Conn.Read(b)
-	return n, err
+	hangConn       chan struct{}
+	isGreetingDone *atomic.Bool
```
Let's name this "`startHanging`" (or similar)?

dfawley commented:
I'd really like to wrap this up before we cut the release branch.  Will that be possible?

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2753,24 +2754,25 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
-		case <-hc.hangConn:
-		case <-timer.C:
-		}
+	if n == goAwayFrameSize { // hang the conn after the goAway is received
```
Why did you remove the atomic?

This is worse, since you're just assuming that: 1. Nothing else written will be the size of the GOAWAY frame, and 2. the GOAWAY will be written to the connection, by itself, in a single operation.

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2754,25 +2755,18 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
 func (hc *hangingConn) Write(b []byte) (n int, err error) {
 	n, err = hc.Conn.Write(b)
-	if n == goAwayFrameSize { // hang the conn after the goAway is received
+	if hc.startHanging.Load() == true {
```
`if x == true` --> `if x`, please

aranjans replied:
done

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2747,70 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	}()
+
+	// Create the server set up.
+	connectCtx, cancel := context.WithDeadline(context.Background(), time.Now().Add(2*time.Second))
```
`context.WithTimeout` instead please

aranjans replied:
done

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2747,70 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+
+	isGreetingDone.Store(true)
+	ct.Close(errors.New("manually closed by client"))
+	close(hangConn)
```
Please defer this above instead.

aranjans replied:
Updated

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2747,70 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+	// Create client transport with custom dialer
+	ct, connErr := NewClientTransport(connectCtx, context.Background(), addr, copts, func(GoAwayReason) {})
+	if connErr != nil {
+		cancel() // Do not cancel in success path.
```
Just defer this after the context is created instead of calling manually and deferring below..

dfawley commented on internal/transport/transport_test.go:
```diff
@@ -2746,3 +2747,70 @@ func (s) TestClientSendsAGoAwayFrame(t *testing.T) {
+		t.Fatalf("failed to create transport: %v", connErr)
+	}
+	defer cancel()
+	defer server.stop()
```
This should be deferred immediately after creation.

aranjans replied:
Updated
